### PR TITLE
[graph_trainer] Preserve mutation dataflow during activation rematerialization

### LINKED ISSUE #4688: [graph_trainer] SAR omits in-place writes to recomputed buffers, producing incorrect gradients
### Bug description

`selective_activation_remat_pass` can silently produce incorrect gradients when a recomputed buffer is modified by a schema-correct in-place operator that returns no value.

The pass follows dataflow producer edges. For a functional `autograd.Function` implemented as `clone(input) -> in-place custom op on the clone -> return clone`, the write has no dataflow output. Recomputing the clone without replaying the write gives backward the pre-mutation value.

The CPU reproduction below computes `loss = (x + 1)^2` at `x = 2`:

- Before the pass: loss = 9, gradient = 6 (correct).
- After the pass: loss = 9, gradient = 4 (incorrect).
- The transformed graph passes `graph.lint()` and executes without an error.

This is a pass-level reproduction using an actual autograd-generated graph, not a full GraphTrainer training launch. Plain `make_fx` does not install GraphTrainer's backward tagging in the tested environment, so the snippet explicitly supplies the backward/recompute metadata expected by the pass. No NPU, CANN, or model assets are required.

Run this script with PyTorch and TorchTitan installed:

```python
"""CPU autograd + make_fx repro of upstream SAR omission; no NPU required."""
import torch
import torch._functorch._activation_offloading.offload_ops
from torch.fx.experimental.proxy_tensor import make_fx
from torch.utils.checkpoint import CheckpointPolicy
from torchtitan.experiments.graph_trainer.selective_activation_remat import selective_activation_remat_pass

lib = torch.library.Library('sar_autograd_review', 'DEF')
lib.define('increment_(Tensor(a!) x) -> ()')
lib.impl('increment_', lambda x: (x.add_(1), None)[1], 'CPU')
lib.impl('increment_', lambda x: None, 'Meta')

class AddOne(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = x.clone()
        torch.ops.sar_autograd_review.increment_(y)
        return y

    @staticmethod
    def backward(ctx, grad):
        return grad

def joint(x):
    y = AddOne.apply(x)
    loss = y.square().sum()
    grad, = torch.autograd.grad(loss, x)
    return loss, grad

x = torch.tensor([2.], requires_grad=True)
gm = make_fx(joint)(x)
# Plain make_fx does not install GraphTrainer's backward tagging in this
# environment. In this deliberately simple graph, everything after loss.sum
# (except output) is autograd-generated backward. Supply the pass metadata
# explicitly, just as in a pass-level regression test.
in_backward = False
for node in gm.graph.nodes:
    if in_backward and node.op != 'output':
        node.meta['autograd_backward'] = True
    if node.target == torch.ops.aten.sum.default:
        in_backward = True
    if node.target == torch.ops.aten.clone.default and not node.meta.get('autograd_backward'):
        node.meta['recompute'] = CheckpointPolicy.MUST_RECOMPUTE

print('before', gm(x))
print('backward-tagged nodes', sum(bool(n.meta.get('autograd_backward')) for n in gm.graph.nodes))
selective_activation_remat_pass(gm)
gm.graph.lint()
print('after SAR', gm(x))
```

Expected behavior: support rematerialization of functional operator wrappers implemented with a private clone followed by a schema-correct in-place kernel. Recomputing the value should preserve the writes required to reconstruct it, so both the forward result and gradients match execution without the rematerialization pass. In this reproduction, the expected loss is 9 and the expected gradient is 6.

We would like GraphTrainer to support this pattern: the wrapper is functional at its public interface, while its implementation uses an in-place kernel on a newly allocated clone. This allows us to use fused operators while retaining activation recomputation. We welcome maintainers' guidance on the appropriate implementation and support scope, including whether to functionalize the graph before SAR or make replay account for the relevant storage dependencies. The proposed direction and eventual support policy are for maintainers to decide; our requested outcome is correct support for this use case.

This was originally observed while using GraphTrainer with a functional partial-RoPE wrapper around an in-place kernel. The CPU example isolates the issue from that hardware/operator implementation. A local experiment that replays the missing write restores the gradient to 6, but a general repair also needs to account for aliases, write ordering, and other write targets. We are reporting the issue first rather than proposing that experiment as a complete fix.

### Versions

- Reproduced on CPU with Python 3.12.10 and `torch==2.14.0+cpu`, on Linux aarch64.
- Rechecked upstream main on September 15, 2026: `f5c97dc0e2d6cca72313004aa853a196ca02ab49` (commit timestamp: 2026-09-15 01:58:39 UTC). Downloaded and directly executed `selective_activation_remat.py` from that exact commit against the CPU reproduction: loss remained 9 and the gradient changed from 6 to 4. The imported `_is_backward_node` and `_get_module_fqn` helpers were also checked against that commit and match. This confirms the defect in the current main SAR pass; it is not a full latest-checkout GraphTrainer training run.
- Upstream source: https://github.com/pytorch/torchtitan/blob/f5c97dc0e2d6cca72313004aa853a196ca02ab49/torchtitan/experiments/graph_trainer/selective_activation_remat.py
- Latest PyTorch/torchao nightly: not independently tested. The reproduction does not invoke torchao operations.
- TOML/config overrides: none; this is a standalone CPU pass-level reproduction.

### PR DESCRIPTION
**Pull request by ZenAlexa**
## Summary

Fixes #4688.

I functionalize schema-declared forward mutations before CPU offload and selective activation rematerialization. This gives SAR explicit dataflow for the values produced by in-place custom operators, including zero-return operators used inside functional autograd wrappers. The issue reproduction now produces loss `9` and gradient `6`; the current main SAR produces loss `9` and gradient `4`.

`_FusedMLAQ.forward` provides an existing mutation consumer for compatibility coverage. Its `fused_mla_q_rope_` operator declares `torch.Tag.inplace` and returns `q`; the baseline already preserves its gradients in the exercised training loop. The zero-return custom operator in #4688 reproduces the missing-write defect. This patch places the repair in the GraphTrainer pass pipeline and uses PyTorch's existing functionalization support.

The pass preserves backward and memory-policy metadata, keeps `DeviceMesh` inputs in the existing graph ABI, and represents ordered effects with internal tokens. FakeTensor tracing preserves real input state during compilation. Runtime input mutations propagate through the functionalized graph. Graphs enter this pass when they contain both recompute annotations and a forward mutable operator.

The existing SAR behavior test now covers private buffers, views, strided views, aliased targets, ordered effects, sequential writes, write-only operators, multiple write targets, and persistent state across repeated calls.

## Validation

The base is `ab1e796219c9c63523e57f2b4d94b0b389aa839a`. Its source tree matches the validated `30af8876d` tree. I ran validation on macOS arm64 CPU with Python 3.12 and PyTorch `2.15.0.dev20260915`.

- Existing memory-policy and SAR coverage: 35 tests and 12 existing subtests passed. The nine mutation scenarios each run three consecutive executions with an explicit `DeviceMesh` input.
- GraphTrainer `minimal_fx_tracer` → memory policy → functionalization → SAR → `run_traced`: static, dynamic, and unbacked shapes with RNG enabled and disabled, ten SGD steps per configuration. Loss, gradients, final parameters, and per-step RNG state match the reference bitwise. Symbolic expressions and `ShapeEnv` identity are preserved.
- Static fullgraph Inductor execution: two ten-step configurations with RNG enabled and disabled match the same-backend reference bitwise.
- Formatting, flake8, AST, license, docstring, spelling, and link checks passed. Pyrefly reports the same 12 environment and existing core errors on the baseline and this patch; the project's configuration excludes `experiments` from that check.

Dynamic fullgraph compilation hits `aten.sym_numel.default` in Dynamo on the unmodified baseline. Dynamic FX replay is covered above.

On one H100 80GB with PyTorch `2.15.0.dev20260911+cu130`, the zero-return CUDA reproduction produces loss `9` and gradient `4` on the baseline, and loss `9` and gradient `6` with this patch.

I also traced the real fused MLA Q operator through GraphTrainer's memory policy and SAR, then ran SGD with momentum for ten steps in FP32 and ten steps in BF16. Loss, gradients, parameters, RoPE buffers, and optimizer momentum match the eager reference bitwise at every step on both revisions. This covers the fused Q training path; full-model training throughput remains outside these measurements.

The existing fused MLA behavior file passes on both revisions on this H100 (10 tests each). The candidate's existing memory-policy and SAR coverage passes with 35 tests and 12 subtests.

### CODE REVIEW

**Comment by aditvenk** [comment]
@claude review

**Comment by qqq6op** [comment]
Possible EP-overlap composability issue: functionalization rewrites MoE's
`tokens_per_expert_E.add_(delta)` into `updated = add(buffer, delta)`, followed
by an outer `copy_(buffer, updated)`. Graph EP chunking then creates two pure
adds that both read the original buffer, while the `(E,)` result has no
batch/sequence chunk dimension from which the full value can be materialized.

When this pass is combined with `--compile.ep_overlap.enabled`, graph chunking
may therefore fail with:

`Chunk pass cannot materialize live-out without chunk dimension ... full users ['copy_...']`

File: torchtitan/experiments/graph_trainer/selective_activation_remat.py [hunk 1]
```diff
@@ -98,6 +106,51 @@ def _collect_backward_regions(
...
+    )
+    fake_inputs = tuple(
+        fake_mode.from_tensor(value) if isinstance(value, torch.Tensor) else value
+        for value in example_inputs
+    )
+    with fake_mode, preserve_node_meta():
+        functional_gm = make_fx(
+            dispatch_functionalize(functionalize, mode, propagate_input_mutations=True),
+        )(*fake_inputs)
+    functional_gm.meta.update(gm.meta)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/selective_activation_remat.py:150]: functionalization could have unexpected side effect globally. 

Although this is a general fix, but I we don't have functional guarantee for the fx graph we are working with.  

If I want functional guarantee, I could have enable functionalization during minimal_fx_trace. 

Let's explore a more localized fix. 

e.g. https://github.com/pytorch/torchtitan/pull/4454 introduced a functional variant of the fused_q_projection op.

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]

**Comment by SherlockNoMad** [comment]
@ZenAlexa 

Hi HKUSTer, alumni here. 
Could you share a bit how are you using Titan GraphTrainer? and why are you using Graph mode? 

This is a new growing community, and I would like to get to know our user a bit.

**Comment by ZenAlexa** [comment]
Thanks for the detailed review. I use GraphTrainer to exercise compile-time graph passes for activation rematerialization, CPU offload, and expert-parallel overlap on the model training path.

The metadata-survival concern is valid. The current numerical cases do not establish that rematerialization and offload remain active after the functionalization trace. I am narrowing the change around the mutation region, adding checks for `recompute`, `autograd_backward`, and `_recomputed` nodes after the pass, and reproducing the EP-overlap composition case before posting the next head.

**Comment by ZenAlexa** [comment]
> @ZenAlexa
> 
> Hi HKUSTer, alumni here. Could you share a bit how are you using Titan GraphTrainer? and why are you using Graph mode?
> 
> This is a new growing community, and I would like to get to know our user a bit.

Hey, great to meet a fellow HKUSTer here! :) 

I've been exploring GraphTrainer to understand how rematerialization, offloading, and expert-parallel overlap fit together in a training step. 

My usage so far has been small training runs and focused experiments. I'm enjoying getting into this part of Titan and seeing where the community takes it~

**Comment by ZenAlexa** [comment]
That makes sense, I'll close this approach and keep the fix at the operator boundary. I reproduced the EP chunking failure with the MoE counter update, so the full-graph pass reaches beyond the mutation we wanted to fix. Recompute metadata does survive on the tested nightly; #4688 remains open with the original reproducer, and #4454 already covers the functional Q operator you pointed to.
